handsontable/src/core.js

```diff
@@ -2018,110 +2019,114 @@ export default function Core(rootElement, userSettings, rootInstanceSymbol = fal
-          }
+        grid.adjustRowsAndCols();
 
-          data.push(row);
+        if (firstRun) {
+          firstRun = [null, 'loadData'];
         }
-      }
-
-    } else {
-      throw new Error(`loadData only accepts array of objects or array of arrays (${typeof data} given)`);
-    }
-
-    if (Array.isArray(data[0])) {
-      instance.dataType = 'array';
-    }
-
-    tableMeta.data = data;
-
-    datamap.dataSource = data;
-    dataSource.data = data;
-    dataSource.dataType = instance.dataType;
-    dataSource.colToProp = datamap.colToProp.bind(datamap);
-    dataSource.propToCol = datamap.propToCol.bind(datamap);
-    dataSource.countCachedColumns = datamap.countCachedColumns.bind(datamap);
-
-    metaManager.clearCellsCache();
-    instance.initIndexMappers();
-
-    grid.adjustRowsAndCols();
+      }, {
+        hotInstance: instance,
+        dataMap: datamap,
+        dataSource,
+        internalSource,
+        source,
+        firstRun
+      });
+  };
 
-    instance.runHooks('afterLoadData', data, firstRun, source);
+  /**
+   * Replaces the dataset with a new one. Unlike `setData` and the `loadData` methods, it doesn't reset the
+   * state of the table.
+   *
+   * @memberof Core#
+   * @function updateData
+   * @since 11.1.0
+   * @param {Array} data Array of arrays or array of objects containing data.
+   * @param {string} [source] Source of the `updateData` call.
+   * @fires Hooks#beforeUpdateData
+   * @fires Hooks#afterUpdateData
+   * @fires Hooks#afterChange
+   */
+  this.updateData = function(data, source) {
+    replaceData(
+      data,
+      (newDataMap) => {
+        datamap = newDataMap;
+      },
+      (newDataMap) => {
+        datamap = newDataMap;
+
+        instance.columnIndexMapper.fitToLength(this.getInitialColumnCount());
+        instance.rowIndexMapper.fitToLength(this.countSourceRows());
+
+        grid.adjustRowsAndCols();
+      }, {
+        hotInstance: instance,
+        dataMap: datamap,
+        dataSource,
+        internalSource: 'updateData',
+        source,
+        firstRun
+      });
+  };
 
-    if (firstRun) {
-      firstRun = [null, 'loadData'];
-    } else {
-      instance.runHooks('afterChange', null, 'loadData');
-      instance.render();
-    }
+  /**
+   * Loads new data to Handsontable. Loading new data resets the cell meta.
+   * Since 8.0.0 loading new data also resets states corresponding to rows and columns
+   * (for example, row/column sequence, column width, row height, frozen columns etc.).
+   *
+   * @memberof Core#
+   * @function loadData
+   * @param {Array} data Array of arrays or array of objects containing data.
+   * @param {string} [source] Source of the loadData call.
+   * @fires Hooks#beforeLoadData
+   * @fires Hooks#beforeSetData
+   * @fires Hooks#afterLoadData
+   * @fires Hooks#afterSetData
+   * @fires Hooks#afterChange
+   */
+  this.loadData = function(data, source) {
+    this.setData(data, source, 'loadData');
```

Comment:
I think that the 3rd argument is still problematic here. As the [comment](https://github.com/handsontable/handsontable/pull/8972#discussion_r757325482) describes I can invoke `hot.setData([[1, 2, 3]], 'edit', 'columnCollapse')` in result the `beforeColumnCollapse` hook is called 🙀

I would be in favor of copy&paste the whole `replaceData` call from `setData` and paste into `loadData` with `internalSource: 'loadData'` then the `/* eslint-enable jsdoc/require-param */` rule could be removed.

Reply:
Not ideal, but you're probably right, that it'd be the safest approach.